src/log/log.c

```diff
@@ -1005,6 +1121,91 @@ __log_newfile(WT_SESSION_IMPL *session, bool conn_open, bool *created)
 }
 
 /*
+ * __log_set_version --
+ *	Set version related information under lock.
+ */
+static int
+__log_set_version(WT_SESSION_IMPL *session,
+    uint16_t major, uint16_t minor, uint32_t first_rec, bool downgrade)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_LOG *log;
+
+	conn = S2C(session);
+	log = conn->log;
+
+	log->log_major = major;
+	log->log_minor = minor;
+	log->first_record = first_rec;
+	if (downgrade)
+		FLD_SET(conn->log_flags, WT_CONN_LOG_DOWNGRADED);
+	else
+		FLD_CLR(conn->log_flags, WT_CONN_LOG_DOWNGRADED);
+	if (!F_ISSET(conn, WT_CONN_READONLY))
+		return (__log_prealloc_remove(session));
+	else
+		return (0);
+}
+
+/*
+ * __wt_log_set_version --
+ *	Change the version number in logging.  Will be done with locking.
+ *	We need to force the log file to advance and remove all old
+ *	pre-allocated files.
+ */
+int
+__wt_log_set_version(WT_SESSION_IMPL *session,
+    uint16_t major, uint16_t minor, uint32_t first_rec, bool downgrade,
+    bool live_chg, uint32_t *lognump)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_LOG *log;
+
+	conn = S2C(session);
+	log = conn->log;
+
+	/*
+	 * The steps are:
+	 * - Set up versions and remove files under lock.
+	 * - Set a flag so that the next slot change forces a file change.
+	 * - Force out the slot that is currently active in the current log.
+	 * - Write a log record to force a record into the new log file.
+	 */
+	WT_WITH_SLOT_LOCK(session, log,
+	    ret = __log_set_version(session,
+	    major, minor, first_rec, downgrade));
+	if (!live_chg)
+		return (ret);
+	WT_ERR(ret);
+	/*
+	 * XXX - Any gap problem between setting version above and this flag
+	 * now? Also may want to clear it in log_newfile when it sees it so that
+	 * additional log records don't also cause new log files repeatedly.
```

Comment:
I can't see a problem with a race here - the `force_write` call follows setting the flag, and switching doesn't clear the flag.